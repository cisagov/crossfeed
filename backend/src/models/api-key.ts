import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToOne
} from 'typeorm';
import { User } from './';

@Entity()
export class ApiKey extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @ManyToOne((type) => User, (user) => user.apiKeys, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  user: User;

  @Column({
    type: 'timestamp',
    nullable: true
  })
  lastUsed: Date | null;

  @Column({
    // length based on hex formatted sha256
    length: 64,
    type: 'varchar'
  })
  hashedKey: string;

  @Column({
    length: 4,
    type: 'varchar'
  })
  lastFour: string;
}
