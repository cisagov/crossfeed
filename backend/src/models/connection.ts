import { createConnection, Connection } from 'typeorm';
import {
  Domain,
  Service,
  Vulnerability,
  Scan,
  Organization,
  User,
  Role,
  ScanTask
} from '.';

let connection: Connection | null = null;

const connectDb = async (logging?: boolean) => {
  const connection = createConnection({
    type: 'postgres',
    host: process.env.DB_HOST,
    port: parseInt(process.env.DB_PORT ?? ''),
    username: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_NAME,
    entities: [
      Domain,
      Service,
      Vulnerability,
      Scan,
      Organization,
      User,
      Role,
      ScanTask
    ],
    synchronize: false,
    name: 'default',
    dropSchema: false,
    logging: logging ?? false
  });
  return connection;
};

export const connectToDatabase = async (logging?: boolean) => {
  if (!connection?.isConnected) {
    connection = await connectDb(logging);
  }
  if (typeof jest === 'undefined') {
    console.log('=> DB Connected');
  }
  return connection;
};
