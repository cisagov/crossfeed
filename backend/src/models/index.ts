export * from './domain';
export * from './service';
export * from './connection';
export * from './vulnerability';
export * from './scan';
export * from './organization';
export * from './user';
export * from './role';
export * from './scan-task';
