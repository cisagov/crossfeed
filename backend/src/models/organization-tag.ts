import {
  Entity,
  Index,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToMany,
  JoinTable,
  Column
} from 'typeorm';
import { Organization, Scan } from '.';

@Entity()
@Index(['name'], { unique: true })
export class OrganizationTag extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column()
  name: string;

  /**
   * Organizations that are labeled with this tag
   */
  @ManyToMany((type) => Organization, (organization) => organization.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  @JoinTable()
  organizations: Organization[];

  /**
   * Scans that have this tag enabled, and will run against all tagged organizations
   */
  @ManyToMany((type) => Scan, (scan) => scan.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  scans: Scan[];
}
