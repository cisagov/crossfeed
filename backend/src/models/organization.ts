import {
  Entity,
  Index,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  OneToMany,
  ManyToMany,
  ManyToOne
} from 'typeorm';
import { Domain, Role, Scan, ScanTask } from '.';
import { User } from './user';

@Entity()
@Index(['name'], { unique: true })
export class Organization extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column()
  name: string;

  @Column('varchar', { array: true })
  rootDomains: string[];

  @Column('varchar', { array: true })
  ipBlocks: string[];

  @Column()
  isPassive: boolean;

  @Column({ default: true })
  inviteOnly: boolean;

  @OneToMany((type) => Domain, (domain) => domain.organization, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  domains: Domain[];

  @OneToMany((type) => Role, (role) => role.organization, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  userRoles: Role[];

  @OneToMany((type) => ScanTask, (scanTask) => scanTask.organization, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  scanTasks: ScanTask[];

  @ManyToMany((type) => Scan, (scan) => scan.organizations, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  granularScans: Scan[];

  @ManyToOne((type) => User, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  createdBy: User;
}
