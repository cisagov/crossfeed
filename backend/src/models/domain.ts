import {
  Entity,
  Index,
  Column,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  BeforeInsert,
  ManyToOne
} from 'typeorm';
import { Service } from './service';
import { Organization } from './organization';
import { Vulnerability } from './vulnerability';
import { Scan } from './scan';
import { Webpage } from './webpage';

@Entity()
@Index(['name', 'organization'], { unique: true })
export class Domain extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  /** When this model was last synced with Elasticsearch. */
  @Column({
    type: 'timestamp',
    nullable: true
  })
  syncedAt: Date | null;

  @Column({
    nullable: true
  })
  ip: string;

  /** Associated root domain that led to the discovery of this domain. */
  @Column({
    nullable: true
  })
  fromRootDomain: string;

  @ManyToOne((type) => Scan, {
    onDelete: 'SET NULL',
    onUpdate: 'CASCADE'
  })
  discoveredBy: Scan;

  @Column({
    length: 512
  })
  reverseName: string;

  @Column({
    length: 512
  })
  name: string;

  @OneToMany((type) => Service, (service) => service.domain)
  services: Service[];

  @OneToMany((type) => Vulnerability, (vulnerability) => vulnerability.domain)
  vulnerabilities: Vulnerability[];

  @OneToMany((type) => Webpage, (webpage) => webpage.domain)
  webpages: Service[];

  @ManyToOne((type) => Organization, { onDelete: 'CASCADE', nullable: false })
  organization: Organization;

  @Column({
    length: 512,
    nullable: true,
    type: 'varchar'
  })
  screenshot: string | null;

  @Column({
    nullable: true,
    type: 'varchar'
  })
  country: string | null;

  @Column({
    nullable: true,
    type: 'varchar'
  })
  asn: string | null;

  @Column({
    default: false
  })
  cloudHosted: boolean;

  /** SSL Certificate information  */
  @Column({
    type: 'jsonb',
    nullable: true
  })
  ssl: {
    issuerOrg: string;
    issuerCN: string;
    validFrom: string;
    validTo: string;
    protocol: string;
    altNames: string;
    bits: string;
    fingerprint: string;
  } | null;

  /** Censys Certificates results */
  @Column({
    type: 'jsonb',
    default: {}
  })
  censysCertificatesResults: {
    [x: string]: any;
  };

  @BeforeInsert()
  setLowerCase() {
    this.name = this.name.toLowerCase();
  }

  @BeforeInsert()
  setReverseName() {
    this.reverseName = this.name.split('.').reverse().join('.');
  }
}
