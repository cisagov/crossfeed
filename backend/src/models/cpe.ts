import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToMany,
  BaseEntity,
  Unique
} from 'typeorm';
import { Cve } from './cve';

@Entity()
@Unique(['name', 'version', 'vendor'])
export class Cpe extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  name: string;

  @Column()
  version: string;

  @Column()
  vendor: string;

  @Column()
  lastSeenAt: Date;

  @ManyToMany(() => Cve, (cve) => cve.cpes)
  cves: Cve[];
}
