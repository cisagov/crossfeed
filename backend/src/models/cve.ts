import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToMany,
  BaseEntity,
  JoinTable,
  Unique
} from 'typeorm';
import { Cpe } from './cpe';

//TODO: Refactor column names to camelCase to match the rest of the codebase?
@Entity()
@Unique(['name'])
export class Cve extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ nullable: true })
  name: string;

  @Column({ nullable: true })
  publishedAt: Date;

  @Column({ nullable: true })
  modifiedAt: Date;

  @Column({ nullable: true })
  status: string;

  @Column({ nullable: true })
  description: string;

  @Column({ nullable: true })
  cvssV2Source: string;

  @Column({ nullable: true })
  cvssV2Type: string;

  @Column({ nullable: true })
  cvssV2Version: string;

  @Column({ nullable: true })
  cvssV2VectorString: string;

  @Column({ nullable: true })
  cvssV2BaseScore: string;

  @Column({ nullable: true })
  cvssV2BaseSeverity: string;

  @Column({ nullable: true })
  cvssV2ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV2ImpactScore: string;

  @Column({ nullable: true })
  cvssV3Source: string;

  @Column({ nullable: true })
  cvssV3Type: string;

  @Column({ nullable: true })
  cvssV3Version: string;

  @Column({ nullable: true })
  cvssV3VectorString: string;

  @Column({ nullable: true })
  cvssV3BaseScore: string;

  @Column({ nullable: true })
  cvssV3BaseSeverity: string;

  @Column({ nullable: true })
  cvssV3ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV3ImpactScore: string;

  @Column({ nullable: true })
  cvssV4Source: string;

  @Column({ nullable: true })
  cvssV4Type: string;

  @Column({ nullable: true })
  cvssV4Version: string;

  @Column({ nullable: true })
  cvssV4VectorString: string;

  @Column({ nullable: true })
  cvssV4BaseScore: string;

  @Column({ nullable: true })
  cvssV4BaseSeverity: string;

  @Column({ nullable: true })
  cvssV4ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV4ImpactScore: string;

  @Column('simple-array', { nullable: true })
  weaknesses: string[];

  @Column('simple-array', { nullable: true })
  references: string[];

  @ManyToMany(() => Cpe, (cpe) => cpe.cves, {
    cascade: true
  })
  @JoinTable()
  cpes: Cpe[];
}
