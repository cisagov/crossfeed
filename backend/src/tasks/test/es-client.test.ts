import { connectToDatabase, Domain, Webpage } from '../../models';
import ESClient from '../es-client';

const bulk = jest.fn(() => ({}));
jest.mock('@elastic/elasticsearch', () => {
  class Client {
    helpers = {
      bulk
    };
  }
  return { Client };
});

let domain;
let webpage;

beforeAll(async () => {
  await connectToDatabase();
  domain = Domain.create({
    name: 'first_file_testdomain5',
    ip: '45.79.207.117'
  });
  webpage = Webpage.create({
    domain,
    url: 'https://cisa.gov/123',
    status: 200,
    updatedAt: new Date('9999-08-23T03:36:57.231Z')
  });
});

describe('updateDomains', () => {
  test('test', async () => {
    const client = new ESClient();
    await client.updateDomains(Array(1).fill(domain));
    expect(bulk).toBeCalledTimes(1);
    expect((bulk.mock.calls[0] as any)[0].datasource.length).toEqual(1);
  });
});

describe('updateWebpages', () => {
  test('test', async () => {
    const client = new ESClient();
    await client.updateWebpages([
      {
        webpage_id: 'webpage_id',
        webpage_createdAt: new Date('2020-10-17T22:21:17Z'),
        webpage_updatedAt: new Date('2020-10-17T22:21:17Z'),
        webpage_syncedAt: new Date('2020-10-17T22:21:17Z'),
        webpage_lastSeen: new Date('2020-10-17T22:21:17Z'),
        webpage_url: 'url',
        webpage_status: 200,
        webpage_domainId: 'domainId',
        webpage_discoveredById: 'webpage_discoveredById',
        webpage_responseSize: 5000,
        webpage_headers: [{ name: 'a', value: 'b' }],
        webpage_body: 'test body'
      }
    ]);
    expect(bulk).toBeCalledTimes(1);
    expect((bulk.mock.calls[0] as any)[0].datasource.length).toEqual(1);
    expect((bulk.mock.calls[0] as any)[0].datasource).toMatchSnapshot();
  });
});
