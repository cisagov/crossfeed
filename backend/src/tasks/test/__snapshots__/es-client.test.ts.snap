// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateWebpages test 1`] = `
Array [
  Object {
    "parent_join": Object {
      "name": "webpage",
      "parent": "domainId",
    },
    "suggest": Array [
      Object {
        "input": "url",
        "weight": 1,
      },
    ],
    "webpage_body": "test body",
    "webpage_createdAt": 2020-10-17T22:21:17.000Z,
    "webpage_discoveredById": "webpage_discoveredById",
    "webpage_domainId": "domainId",
    "webpage_headers": Array [
      Object {
        "name": "a",
        "value": "b",
      },
    ],
    "webpage_id": "webpage_id",
    "webpage_lastSeen": 2020-10-17T22:21:17.000Z,
    "webpage_responseSize": 5000,
    "webpage_status": 200,
    "webpage_syncedAt": 2020-10-17T22:21:17.000Z,
    "webpage_updatedAt": 2020-10-17T22:21:17.000Z,
    "webpage_url": "url",
  },
]
`;
